[processor/tailsampling] Add first policy match decision to tailsampling processor

stollero opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
The policies are evaluated in order, and the first policy that matches will be used to determine the sample rate.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
#36795

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Unit-Test added

<!--Describe the documentation added.-->
#### Documentation
✅ new config added to README

<!--Please delete paragraphs that you did not use before submitting.-->

stollero committed:
[processor/tailsampling] Add first policy match decision to tailsampling processor

atoulme commented:
Please address conflicts. @portertech please review

portertech commented:
@stollero I think we need to square this with the newly added `drop` policy type. If a user were to firstly configure their drop policies and make use of first match, it will currently behave as expected. If a drop policy is configured after a policy that matches, it will NOT behave as expected. We can consider manipulating policy order if drop policies are configured. Thoughts?

portertech commented:
@stollero let's begin w/ manipulating the policy order, moving `drop` policies to the top 👍 Should be able to wire it into `loadSamplingPolicy()`. This doesn't break existing behaviour and as you pointed out, likely has performance advantages regardless of first match.

stollero commented:
@portertech sounds great! I will implement the change 👍

stollero committed:
[processor/tailsampling] Always evaluate drop policies first

- performance optimization
- drop policy takes precedence over all other, even first policy match

portertech commented on processor/tailsamplingprocessor/README.md:
```diff
@@ -60,6 +60,7 @@ The following configuration options can also be modified:
   - `non_sampled_cache_size` (default = 0) Configures amount of trace IDs to be kept in an LRU cache,
     persisting the "drop" decisions for traces that may have already been released from memory.
     By default, the size is 0 and the cache is inactive.
+- `sample_on_first_match`: Make decision as soon as a policy matches
```
nit, indentation.

stollero replied:
If I indent it, people will think it is a additional setting for the `decision_cache` :)

portertech replied:
My mistake 👍

portertech commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -430,6 +439,11 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pcommon.TraceID, trace *sa
 		}
+		// If sampleOnFirstMatch is enabled, make decision as soon as a policy matches
+		if tsp.sampleOnFirstMatch && decision == sampling.Sampled {
+			finalDecision = decision
```
Is this necessary? It gets set again in the switch.

stollero replied:
You are right. I fixed it

stollero committed:
[processor/tailsamping] Remove unnecessary statement, it already gets set inside the switch

portertech commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -285,14 +288,20 @@ func (tsp *tailSamplingSpanProcessor) loadSamplingPolicy(cfgs []PolicyCfg) error
+		}
+	}
+	// Dropped decision takes precedence over all others, therefore we evaluate them first.
+	tsp.policies = append(dropPolicies, policies...)
```
Lint issue

stollero replied:
My bad, should have used `Concat` instead. Fixed it 👍

stollero committed:
[processor/tailsampling] use concat to create a new combined slice instead of append

which could be misread of result in undefined behaviour
